### PR TITLE
ngrok: switch to mutable builder pattern

### LINKED ISSUE #95: Builder ergonomics
Hi! :wave: 

First, let me thank you for the great crate. It was a breeze to add [initial support for ngrok](https://github.com/Devolutions/devolutions-gateway/commit/711164010a6660f2946efb99a24ccb1a4cd47ba1) in our product, [Devolutions Gateway](https://github.com/Devolutions/devolutions-gateway/).

However, I noticed all the builders are "consuming". This means, we can’t write straightforward code such as:
```rust
let mut builder = Session::builder()
    .authtoken(…)
    .tcp_endpoint()
    .remote_addr(…)
    .forwards_to(…);

if let Some(metadata) = metadata {
    builder.metadata(metadata);
}

for deny_cidr in deny_cidrs {
    builder.deny_cidr(deny_cidr);
}

let tunnel = builder.listen().await?;
```

This is really important when the configuration must be dynamic. See how I had to resort to macro tricks to get something somewhat "clean": https://github.com/Devolutions/devolutions-gateway/pull/442

Here is an excerpt:
```rust
macro_rules! builder_call_opt {
    ($builder:ident . $method:ident ( $ngrok_option:expr ) ) => {{
        if let Some(option) = $ngrok_option {
            $builder.$method(option)
        } else {
            $builder
        }
    }};
}

macro_rules! builder_call_vec {
    ($builder:ident . $method:ident ( $ngrok_option:expr ) ) => {{
        let mut builder = $builder;
        let mut iter = $ngrok_option;
        loop {
            builder = match iter.next() {
                Some(item) => builder.$method(item),
                None => break builder,
            };
        }
    }};
    ($ngrok_option:expr, $builder:ident . $method:ident ( $( $( & )? $field:ident ),+ ) ) => {{
        let mut builder = $builder;
        let mut iter = $ngrok_option.iter();
        loop {
            builder = match iter.next() {
                Some(item) => builder.$method($( & item . $field ),+),
                None => break builder,
            };
        }
    }};
}

macro_rules! builder_call_flag {
    ($builder:ident . $method:ident ( $ngrok_option:expr ) ) => {{
        match $ngrok_option {
            Some(option) if option => $builder.$method(),
            _ => $builder,
        }
    }};
}

let builder = Session::builder().authtoken(&conf.authtoken);
let builder = builder_call_opt!(builder.heartbeat_interval(conf.heartbeat_interval));
let builder = builder_call_opt!(builder.heartbeat_tolerance(conf.heartbeat_tolerance));
let builder = builder_call_opt!(builder.metadata(&conf.metadata));
let builder = builder_call_opt!(builder.server_addr(&conf.server_addr));

let session = builder.connect().await?;

let builder = session.http_endpoint().domain(&http_conf.domain).forwards_to(hostname);
let builder = builder_call_opt!(builder.metadata(&http_conf.metadata));
let builder = builder_call_vec!(http_conf.basic_auth, builder.basic_auth(username, password));
let builder = builder_call_opt!(builder.circuit_breaker(http_conf.circuit_breaker));
let builder = builder_call_flag!(builder.compression(http_conf.compression));
let builder = builder_call_vec!(builder.allow_cidr(http_conf.allow_cidrs.iter()));
let builder = builder_call_vec!(builder.deny_cidr(http_conf.deny_cidrs.iter()));
let builder = builder_call_opt!(builder.proxy_proto(http_conf.proxy_proto.map(ProxyProto::from)));
let builder = builder_call_vec!(builder.scheme(
    http_conf
        .schemes
        .iter()
        .map(|s| Scheme::from_str(s).unwrap_or(Scheme::HTTPS))
));

let tunnel = builder.listen().await?;
```

For reference, I had a similar situation in our [IronRDP](https://github.com/Devolutions/IronRDP) WASM module: [SessionBuilder](https://github.com/Devolutions/IronRDP/blob/124c3aa2512964a327af75d6c93d8f7f43532952/crates/ironrdp-web/src/session.rs#L30). In order to interface with JavaScript in a clean way, the builder methods must not take ownership of `self`. Using a `RefCell` internally is often the best solution. (An additional requirement, that ngrok-rs can ignore, was for the builder to be cheap to clone, hence the `Rc`.)

Rust official [Style Guidelines book](https://doc.rust-lang.org/1.0.0/style/) is also a good read: https://doc.rust-lang.org/1.0.0/style/ownership/builders.html#non-consuming-builders-(preferred)

### PR DESCRIPTION
**Pull request by jrobsonchase**
Resolves #95

This also opens the way to allow builder methods to return `Results`, as
the builder won't be consumed in the event of bad input. Then, we can
push some validation to build time rather than connect time.

Combining this with a few other breaking changes, namely:
* No more `Error` stutter
* Validate the server addr at config time
* Validate the heartbeat config at config time
* Remove the server addr/heartbeat errors from the connect error enum
* Split the server addr into host/port when passing them to the connector

### CODE REVIEW

File: ngrok/src/session.rs [hunk 1]
```diff
@@ -198,24 +197,18 @@ where
...
-        .unwrap_or(443);
-    let conn = tokio::net::TcpStream::connect(&(host, port))
+    let conn = tokio::net::TcpStream::connect(&(host.as_str(), port))
         .await
         .map_err(ConnectError::Tcp)?
         .compat();
 
     let tls_conn = async_rustls::TlsConnector::from(tls_config)
-        .connect(rustls::ServerName::try_from(host)?, conn)
+        .connect(rustls::ServerName::try_from(host.as_str()).unwrap(), conn)
```
Comment (bobzilladev) [inline ngrok/src/session.rs:211]: Why the change to `unwrap` here?

**Review by bobzilladev — APPROVED** [review]
lgtm. examples still happy. i'm a bit scared on how much is going to need to change in the downstream ffi sdk's.

File: ngrok/src/session.rs (hunk 1 above)
Comment (jrobsonchase) [reply ngrok/src/session.rs:211]: It's getting validated in the `server_addr` builder method. I really wanted to pass a pre-validated type, but for some reason none of the ones available allowed going from validated type back to a string, which is needed for the TCP connection :confused: 

I can add a comment to ward off future readers here

Comment (jrobsonchase) [reply ngrok/src/session.rs]: Changed the unwrap to a clarifying `expect`
